shared/ui/controlled/controlled-text-field.tsx

```diff
@@ -0,0 +1,42 @@
+import { Control, FieldPath, FieldValues, useController } from 'react-hook-form';
+
+import { assertUnreachable } from '@/shared/helpers/assertUnreachable';
+import { TextField, TextFieldProps } from '@/shared/ui/textField';
+
+type ControllerProps<TFieldValues extends FieldValues> = {
+  control: Control<TFieldValues>;
+  name: FieldPath<TFieldValues>;
+};
+type FilteredTextFieldProps<T extends TextFieldProps> = Omit<
+  T,
+  'error' | 'onBlur' | 'onChange' | 'value'
+>;
+
+// prettier-ignore
+export type ControlledTextFieldProps<TFieldValues extends FieldValues> =
+  | (ControllerProps<TFieldValues> &
+      FilteredTextFieldProps<Extract<TextFieldProps, { as: 'textarea' }>>)
+  | (ControllerProps<TFieldValues> &
+      FilteredTextFieldProps<Extract<TextFieldProps, { as?: 'input' }>>);
+
+export const ControlledTextField = <TFieldValues extends FieldValues = FieldValues>({
+  control,
+  disabled,
+  name,
+  ...props
+}: ControlledTextFieldProps<TFieldValues>) => {
+  const {
+    field,
+    fieldState: { error },
+  } = useController({ control, disabled, name });
+
+  if (!props.as || props.as === 'input') {
+    return <TextField {...props} error={error?.message} {...field} />;
+  }
+
+  if (props.as === 'textarea') {
+    return <TextField {...props} error={error?.message} {...field} />;
+  }
+
+  assertUnreachable(props.as);
```

Comment:
`assertUnreachable` asserts the exhaustive nature of the passed value. In other words, it will show an error if the passed value can continue to exist after attempting to process all of its potential values.
This function in particular can be useful to ensure exhaustivness of the `switch` block ([example](https://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript?noredirect=1&lq=1)).
#### [More about never type](https://marketsplash.com/tutorials/typescript/typescript-never/)